octree/include/pcl/octree/octree_iterator.h

```diff
@@ -568,6 +568,114 @@ namespace pcl
         std::deque<IteratorState> FIFO_;
       };
 
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    /** \brief @b Octree iterator class
+     * \note Iterator  over all existing nodes at a given depth. It walks across an octree
+     *       in a breadth-first manner.
+     * \ingroup octree
+     * \author Fabien Rozar (fabien.rozar@gmail.com)
+     */
+    template<typename OctreeT>
+    class OctreeFixedDepthIterator : public OctreeBreadthFirstIterator<OctreeT>
+    {
+    public:
+
+      // public typedefs
+      using typename OctreeBreadthFirstIterator<OctreeT>::BranchNode;
+      using typename OctreeBreadthFirstIterator<OctreeT>::LeafNode;
+
+      /** \brief Empty constructor.
+       * \param[in] fixed_depth_arg Depth level during traversal
+       */
+      explicit
+      OctreeFixedDepthIterator (unsigned int fixed_depth_arg = 0);
```

Comment:
What does it mean to have an iterator without a tree?

Reply:
If you take a look at the following test case:
https://github.com/PointCloudLibrary/pcl/blob/754894342dad0edfd8e44fd9e04dca6408033ab7/test/octree/test_octree_iterator.cpp#L149-L164

At line 156, you have an iterator which is instantiated whitout a tree. If this kind of code is allowed, you need a constructor without a tree (even if it seems weird). In my opinion, it's necessary.

Reply:
If I think about it, I never default initialized any of `std::xxx::iterator` types: it's always generated from a `begin()`, `end()` or copy ctor. I'm inclined to say we don't really need it.

@frozar do you invoke default constructor iterators in code you normally write? I'm asking for usual iterators used withing the standard library.

Reply:
If I think about it, I never default initialized any of `std::xxx::iterator` types: it's always generated from a `begin()`, `end()` or copy ctor. I'm inclined to say we don't really need it.

@frozar do you invoke default constructor iterators in code you normally write? I'm asking for usual iterators used withing the standard library.

Reply:
OK, actually iterators in STL are often default constructible. Perhaps the reason is e.g. to be able to have an iterator member field without needing to initialize it right away in the class constructor.

So we can keep it. However, I think it should be parameterless, there is absolutely no use case where you would create an iterator without a tree but with some particular depth.